### PR TITLE
[ruutu] Add hs.fi embeds

### PR DESCRIPTION
**Pull request by tpikonen**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Improvement

---

### Description of your *pull request* and other information

Add Ruutu embeds from hs.fi.

### CODE REVIEW

File: yt_dlp/extractor/ruutu.py [hunk 1]
```diff
@@ -126,14 +126,27 @@ class RuutuIE(InfoExtractor):
...
+        settings = try_call(
+            lambda: json.loads(re.search(
+                r'<script id="__NEXT_DATA__" type="application/json">\s*({.+?})\s*</script>',
+                webpage).group(1), strict=False))
+        if settings:
+            video_ids = list(filter(
+                bool, (traverse_obj(d, ('video', 'sourceId')) for d in traverse_obj(settings,
+                       ('props', 'pageProps', 'page', 'assetData', 'splitBody')))))
+            if video_ids:
+                return list(f'http://www.ruutu.fi/video/{v}' for v in video_ids)
```
Comment (pukkandan) [inline yt_dlp/extractor/ruutu.py:149]: yield instead

File: yt_dlp/extractor/ruutu.py [hunk 2]
```diff
@@ -126,14 +126,27 @@ class RuutuIE(InfoExtractor):
...
+                return [f'http://www.ruutu.fi/video/{video_id}']
+        # hs.fi
+        settings = try_call(
+            lambda: json.loads(re.search(
+                r'<script id="__NEXT_DATA__" type="application/json">\s*({.+?})\s*</script>',
+                webpage).group(1), strict=False))
+        if settings:
+            video_ids = list(filter(
+                bool, (traverse_obj(d, ('video', 'sourceId')) for d in traverse_obj(settings,
+                       ('props', 'pageProps', 'page', 'assetData', 'splitBody')))))
```
Comment (pukkandan) [inline yt_dlp/extractor/ruutu.py:147]: this is unreadable. I can barely tell where each bracket ends. Also, traverse_obj can handle all this filter and loop. Try this: (untested)
```suggestion
            video_ids = traverse_obj(settings, ('props', 'pageProps', 'page', 'assetData', 'splitBody', ..., 'video', 'sourceId'))
```
PS: The `...` is literal

File: yt_dlp/extractor/ruutu.py [hunk 3]
```diff
@@ -126,14 +126,27 @@ class RuutuIE(InfoExtractor):
...
+        if settings:
+            video_id = traverse_obj(settings, (
+                'mediaCrossbowSettings', 'file', 'field_crossbow_video_id', 'und', 0, 'value'))
+            if video_id:
+                return [f'http://www.ruutu.fi/video/{video_id}']
+        # hs.fi
+        settings = try_call(
+            lambda: json.loads(re.search(
+                r'<script id="__NEXT_DATA__" type="application/json">\s*({.+?})\s*</script>',
+                webpage).group(1), strict=False))
```
Comment (pukkandan) [inline yt_dlp/extractor/ruutu.py:143]:
```suggestion
            lambda: json.loads(re.search(
                '(?s)<script[^>]+id=[\'"]__NEXT_DATA__[\'"][^>]*>([^<]+)</script>',
                webpage).group(1), strict=False))
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/ruutu.py (hunk 1 above)
Comment (tpikonen) [reply yt_dlp/extractor/ruutu.py]: Do you mean return a generator (see the first new commit)?

File: yt_dlp/extractor/ruutu.py (hunk 2 above)
Comment (tpikonen) [reply yt_dlp/extractor/ruutu.py]: Cool, thanks. The new form above works fine.

File: yt_dlp/extractor/ruutu.py (hunk 3 above)
Comment (tpikonen) [reply yt_dlp/extractor/ruutu.py]: Thanks, I also added this as is.

**Comment by tpikonen** [comment]
Thanks for the review. I made your suggested changes and added support for a type of page which was not working before, plus a test case.

**Pull request merged**